File: src/tools/packages.ts
import { z } from "zod";
import { type ToolRegistrar, jsonResult } from "./context.js";

export const registerPackageTools: ToolRegistrar = (server, ctx) => {
  server.tool(
    "list_categories",
    "List all categories and their packages as displayed on the webstore. " +
      "Returns the full category tree with nested package summaries.",
    {},
    async () => {
      const data = await ctx.client.getListing();
      return jsonResult(data);
    },
  );

  server.tool(
    "list_packages",
    "List all packages in the store with full details: id, name, price, " +
      "image, description, type, category, and sale info.",
    {},
    async () => {
      const data = await ctx.client.listPackages();
      return jsonResult(data);
    },
  );

  server.tool(
    "get_package",
    "Get full details of a single package by its ID.",
    {
      package_id: z.number().int().describe("Package ID"),
    },
    async ({ package_id }) => {
      const data = await ctx.client.getPackage(package_id);
      return jsonResult(data);
    },
  );

  server.tool(
    "update_package",
    "Update a package's properties. Only provided fields are changed. " +
      "Can toggle disabled state, rename, or change price.",
    {
      package_id: z.number().int().describe("Package ID"),
      disabled: z
        .boolean()
        .optional()
        .describe("Set to true to disable the package on the webstore"),
      name: z.string().optional().describe("New package name"),
      price: z
        .number()
        .min(0)
        .optional()
        .describe("New price in store currency"),
    },
    async ({ package_id, disabled, name, price }) => {
      await ctx.client.updatePackage(package_id, { disabled, name, price });
      return jsonResult({
        ok: true,
        package_id,
        updated: { disabled, name, price },
      });
    },
  );
};

File: src/tebex_mcp/tools/packages.py
"""Package tools."""

from __future__ import annotations

from typing import Annotated, Any

from fastmcp import FastMCP
from pydantic import Field

from tebex_mcp import normalize
from tebex_mcp.client import TebexError
from tebex_mcp.context import ToolContext
from tebex_mcp.tools._common import IDEMPOTENT, READ_ONLY, map_tebex_errors, ok


def register(mcp: FastMCP, ctx: ToolContext) -> None:
    @mcp.tool(
        name="list_categories",
        description="List all categories with nested package summaries.",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def list_categories() -> Any:
        return await ctx.client.get_listing()

    @mcp.tool(
        name="list_packages",
        description="List all packages: id, name, price, type, category, expiry, limits, disabled.",
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def list_packages() -> Any:
        return await ctx.client.list_packages()

    @mcp.tool(
        name="get_package",
        description=(
            "Get a package's full config: price, type, category, expiry, limits, "
            "GUI item, server/requirement rules. Set include_description to also "
            "fetch the storefront view (description, tax-inclusive pricing, media, "
            "options) under a `storefront` key — one extra Headless API call."
        ),
        annotations=READ_ONLY,
    )
    @map_tebex_errors
    async def get_package(
        package_id: Annotated[int, Field(description="Package ID", ge=1)],
        include_description: Annotated[
            bool,
            Field(description="Also fetch the storefront description/pricing/media"),
        ] = False,
    ) -> Any:
        pkg = await ctx.client.get_package(package_id)
        if include_description and isinstance(pkg, dict):
            try:
                storefront = normalize.package_storefront(
                    await ctx.client.get_package_storefront(package_id)
                )
            except TebexError:
                storefront = {"error": "not available on the storefront"}
            pkg = {**pkg, "storefront": storefront}
        return pkg

    @mcp.tool(
        name="update_package",
        description="Update a package: toggle disabled, rename, or change price. Only provided fields are changed.",
        annotations=IDEMPOTENT,
    )
    @map_tebex_errors
    async def update_package(
        package_id: Annotated[int, Field(description="Package ID", ge=1)],
        disabled: Annotated[
            bool | None,
            Field(description="Set to true to disable the package on the webstore"),
        ] = None,
        name: Annotated[str | None, Field(description="New package name")] = None,
        price: Annotated[
            float | None, Field(description="New price in store currency", ge=0)
        ] = None,
    ) -> dict[str, Any]:
        await ctx.client.update_package(
            package_id, disabled=disabled, name=name, price=price
        )
        return ok(
            package_id=package_id,
            updated={"disabled": disabled, "name": name, "price": price},
        )
